src/main/java/com/vmware/vipclient/i18n/messages/api/opt/server/RemotePatternOpt.java

```diff
@@ -4,82 +4,88 @@
+    public void getPatterns(String language, String region, PatternCacheItem cacheItem) {
         logger.debug("Look for pattern from Singleton Service for language [{}], region [{}]!", language, region);
-        String responseStr = "";
-        String i18nScope = VIPCfg.getInstance().getI18nScope();
         HttpRequester httpRequester = VIPCfg.getInstance().getVipService().getHttpRequester();
-        if (i18nScope != null && !"".equalsIgnoreCase(i18nScope)) {
-        	Map<String, Object> response = httpRequester.request(V2URL.getPatternURL(language, region,
-                    httpRequester.getBaseURL()), ConstantsKeys.GET, null);
-        	responseStr = (String) response.get(URLUtils.BODY);
-        }
-        if (null == responseStr || responseStr.equals("")) {
-            return null;
-        } else {
-            Object dataObj = this.getCategoriesFromResponse(responseStr, PatternKeys.CATEGORIES);
-            JSONObject msgObject = null;
-            if (dataObj != null) {
-                msgObject = (JSONObject) dataObj;
+        getPatternsFromRemote(language+"-"+region, V2URL.getPatternURL(language, region,
+                    httpRequester.getBaseURL()), ConstantsKeys.GET, null, cacheItem);
+    }
+
+    private void getPatternsFromRemote(String locale, String url, String method, Object requestData, PatternCacheItem cacheItem) {
+
+        Map<String, Object> response = getResponse(url, method, requestData, cacheItem);
+
+        Integer responseCode = (Integer) response.get(URLUtils.RESPONSE_CODE);
+
+        if (responseCode != null && (responseCode.equals(HttpURLConnection.HTTP_OK) ||
+                responseCode.equals(HttpURLConnection.HTTP_NOT_MODIFIED))) {
+            long timestamp = 0;
+            String etag = null;
+            Long maxAgeMillis = null;
+
+            if (response.get(URLUtils.RESPONSE_TIMESTAMP) != null)
+                timestamp = (long) response.get(URLUtils.RESPONSE_TIMESTAMP);
+            if (response.get(URLUtils.HEADERS) != null)
+                etag = URLUtils.createEtagString((Map<String, List<String>>) response.get(URLUtils.HEADERS));
+            if (response.get(URLUtils.MAX_AGE_MILLIS) != null)
+                maxAgeMillis = (Long) response.get(URLUtils.MAX_AGE_MILLIS);
+
+            if (responseCode.equals(HttpURLConnection.HTTP_OK)) {
+                try {
+                    String responseBody = (String) response.get(URLUtils.BODY);
+                    Map<String, Object> patterns = getPatternsFromResponse(responseBody);
+                    if (patterns != null) {
+                        logger.debug("Find the pattern from Singleton Service for locale [{}].\n", locale);
+                        cacheItem.set(patterns, etag, timestamp, maxAgeMillis);
+                    }else{
+                        logger.debug("Doesn't find the pattern from Singleton Service for locale [{}].\n", locale);
+                        cacheItem.set(etag, timestamp, maxAgeMillis);
```

Comment:
This is for handling non-supported locales, just want to add an empty fetch result to cache, it's the same as you new an empty MessageCacheItem and add it to cache.

Reply:
Same reply as in RemoteLocaleOpt
_"The cacheItem object here was passed a parameter. It could be an object that was retrieved from the cache (not a new cacheItem), updating its timestamp here incorrectly "un-expires" the cacheItem. We want to keep the old timestamp if the fetch failed.

If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, remove this line please."_

Reply:
Same reply as in RemoteLocaleOpt
_"The cacheItem object here was passed a parameter. It could be an object that was retrieved from the cache (not a new cacheItem), updating its timestamp here incorrectly "un-expires" the cacheItem. We want to keep the old timestamp if the fetch failed.

If you noticed in my most recent PR, we have changed the design of caching empty cacheItem objects for non-supported locales. This is because this design won't help in the case of the very first request (it will still be slow). So for L3, we just use the list of supported locales to figure out the next steps.. and in the case of non-supported locale, that would be to return fallback locale messages.

For L2, can we do the same in next PR? For now, remove this line please."_

Reply:
Yeah, I know your optimization for this. That's great!

But currently it's impossible for L2 do the same as L3. Because to judge if the requested locale is supported or not, we need get the supported locale list first like L3, but currently there is no API for getting supported locale list of L2 in Singleton service. Maybe this can't be done in the next PR/in a short time. So for L2 currently I can't resolve the slowness at first request, only can avoid the slowness in later request by adding an empty CacheItem.

You're right the cacheItem object maybe retrieved from the cache, but updating its timestamp happens after cacheItem expires, I think it's correct to update the timestamp to un-expires the cacheItem no matter fetch success or failed, then next fetch will happen after cacheItem expired. But keeping the old timestamp will always make the cacheItem expire and fetch from data source until fetch success which may take up resources.

So I don't agree with you on keeping the old timestamp.